### PR TITLE
feat: EIP-1153 Transient storage opcodes

### PR DESCRIPTION
**Pull request by tynes**
Implement EIP-1153 for the Cancun hardfork

The changes to the specs in this PR I'd like some feedback on, happy to revert things that aren't directly related to the EIP implementation

Need to update the opcode numbers before this PR is ready, they are still on the older values

https://eips.ethereum.org/EIPS/eip-1153

Closes https://github.com/bluealloy/revm/issues/154

### CODE REVIEW

File: crates/interpreter/src/instructions/host.rs [hunk 1]
```diff
@@ -179,6 +179,24 @@ pub fn sstore<SPEC: Spec>(interpreter: &mut Interpreter, host: &mut dyn Host) {
     refund!(interpreter, gas::sstore_refund::<SPEC>(original, old, new));
 }
 
+pub fn tstore<H: Host, SPEC: Spec>(interpreter: &mut Interpreter, host: &mut H) {
+    check_staticcall!(interpreter);
```
Comment (rakita) [inline crates/interpreter/src/instructions/host.rs:183]: Need additional check for hardfork. Both here and for tload

File: crates/interpreter/src/instructions/host.rs [hunk 2]
```diff
@@ -179,6 +179,24 @@ pub fn sstore<SPEC: Spec>(interpreter: &mut Interpreter, host: &mut dyn Host) {
...
+
+    pop!(interpreter, index, value);
+
+    host.tstore(interpreter.contract.address, index, value);
+}
+
+pub fn tload<H: Host, SPEC: Spec>(interpreter: &mut Interpreter, host: &mut H) {
+    gas!(interpreter, gas::WARM_STORAGE_READ_COST);
+
+    pop!(interpreter, index);
```
Comment (rakita) [inline crates/interpreter/src/instructions/host.rs:194]: you can use `pop_top!` here and you set tload value to it.

File: crates/revm/src/evm.rs [hunk 3]
```diff
@@ -196,7 +196,6 @@ pub fn evm_inner<'a, DB: Database, const INSPECT: bool>(
         }
         SpecId::MERGE => create_evm!(MergeSpec, db, env, insp),
         SpecId::SHANGHAI => create_evm!(ShanghaiSpec, db, env, insp),
-        SpecId::CANCUN => create_evm!(LatestSpec, db, env, insp),
-        SpecId::LATEST => create_evm!(LatestSpec, db, env, insp),
+        SpecId::CANCUN | SpecId::LATEST => create_evm!(LatestSpec, db, env, insp),
```
Comment (rakita) [inline crates/revm/src/evm.rs:199]: LATEST should be LatestSpec, Cancun should have its own spec

File: crates/revm/src/journaled_state.rs [hunk 4]
```diff
@@ -657,6 +690,23 @@ impl JournaledState {
         Ok((slot.original_value, present, new, is_cold))
     }
 
+    pub fn tstore(&mut self, address: B160, key: U256, new: U256) {
+        let present = self.tload(address, key);
```
Comment (rakita) [inline crates/revm/src/journaled_state.rs:694]: We are not calling `tstore` anywhere

File: crates/revm/src/journaled_state.rs [hunk 5]
```diff
@@ -393,6 +407,13 @@ impl JournaledState {
                         storage.remove(&key);
                     }
                 }
+                JournalEntry::TransientStorageChange {
+                    address,
+                    key,
+                    had_value,
+                } => {
+                    transient_storage.set(address, key, had_value.unwrap());
```
Comment (rakita) [inline crates/revm/src/journaled_state.rs:415]: If `had_value` is `None` we should remove the entry from transient storage.

File: crates/interpreter/src/interpreter.rs [hunk 6]
```diff
@@ -2,6 +2,7 @@ pub mod analysis;
 mod contract;
 pub(crate) mod memory;
 mod stack;
+pub mod transient;
```
Comment (rakita) [inline crates/interpreter/src/interpreter.rs:5]: Would be good to have full name `transient_storage`

**Comment by tynes** [comment]
Thank you for picking this up @rakita. For test coverage, do you think using the Ethereum state tests are sufficient?

The implementation looks good to me, looks much more idiomatic than before

**Comment by rakita** [comment]
> Thank you for picking this up @rakita. For test coverage, do you think using the Ethereum state tests are sufficient?
> 
> The implementation looks good to me, looks much more idiomatic than before

using eth tests should be sufficient to cover edge cases
